<a href="https://colab.research.google.com/github/valeriaquiroz/Technical-test-Globant/blob/main/Purchase.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Produced by:** Valeria Quiroz Gomez **Date**:September 2021

**Point 2:** https://data.birmingham.gov.uk/dataset/purchase-card-transactions


**At this point only an exploratory analysis of variables is carried out**

In the URL you will find information about the exercise. 

Data Set Information: you have information about purchases by credit card, in this case, a period of time from 2017 is selected, to carry out different data science activities


In [45]:
#Import necessary libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt #plot
import seaborn as sns #plot
from sklearn.model_selection import train_test_split #model
from sklearn.model_selection import cross_val_score #validation
from sklearn.metrics import accuracy_score, precision_score, recall_score #metrics of evaluation

#for model 
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import classification_report, f1_score, accuracy_score, confusion_matrix
from sklearn.metrics import roc_curve, auc, roc_auc_score
from sklearn import metrics 
from sklearn.ensemble import RandomForestClassifier


In [13]:
#read to dataset 
df1 = pd.read_excel("/content/spend-january-2017.xls", index_col=None)
df2 = pd.read_excel("/content/spend-february-2017.xls",index_col=None)
df3 = pd.read_excel("/content/spend-march-2017.xls")
df4 = pd.read_excel("/content/spend-april-2017.xls")
df5 = pd.read_excel("/content/spend-may-2017.xls")
df6 = pd.read_excel("/content/spend-june-2017.xls")
df7 = pd.read_excel("/content/spend-july-2017.xls")
df8 = pd.read_excel("/content/spend-august-2017.xls")
df9 = pd.read_excel("/content/spend-sept-2017.xls")
df10 = pd.read_excel("/content/spend-october-2017.xls")
df11 = pd.read_excel("/content/spend-november-2017.xls")
df1#see to dataFrame 

TRANS DATE TRANS VAT DESC  ...  TRANS CAC CODE 3          Directorate
0    2017-01-03             VR  ...               A00  Adult & Communities
1    2017-01-24             VR  ...               A00  Adult & Communities
2    2017-02-01             VR  ...               A00  Adult & Communities
3    2017-01-04             VR  ...               A00  Adult & Communities
4    2017-01-05             VR  ...               A00  Adult & Communities
...         ...            ...  ...               ...                  ...
5222 2017-01-31             VR  ...               A00       Local Services
5223 2017-02-02             VR  ...               A00       Local Services
5224 2017-01-19             VR  ...               A00       Local Services
5225 2017-01-20             VZ  ...               A00       Local Services
5226 2017-01-20             VZ  ...               A00       Local Services

[5227 rows x 14 columns]

The data are **concatenated** to have a single data set and subsequently perform the respective analysis

In [21]:
data=pd.concat([df1, df2, df3, df4, df5,df6,df7,df8,df9,df10,df11], ignore_index=True)
data

TRANS DATE TRANS VAT DESC  ...  BILLING CUR CODE.1 Directorates
0     2017-01-03             VR  ...                 NaN          NaN
1     2017-01-24             VR  ...                 NaN          NaN
2     2017-02-01             VR  ...                 NaN          NaN
3     2017-01-04             VR  ...                 NaN          NaN
4     2017-01-05             VR  ...                 NaN          NaN
...          ...            ...  ...                 ...          ...
48584 2017-11-23             VR  ...                 NaN          NaN
48585 2017-11-17             VR  ...                 NaN          NaN
48586 2017-11-17             VR  ...                 NaN          NaN
48587 2017-11-24             VR  ...                 NaN          NaN
48588 2017-11-24             VZ  ...                 NaN          NaN

[48589 rows x 17 columns]


The result is a data set with 48589 rows and 17 columns


**initial statistics and explore data set**

In [22]:
data.describe()

ORIGINAL GROSS AMT  BILLING GROSS AMT  TRANS TAX AMT
count        48589.000000       48589.000000    4598.000000
mean           215.459431         200.594515      34.268212
std           2893.488479        2488.629938     149.519355
min        -486980.090000     -486980.090000    -440.650000
25%             13.260000          13.250000       0.000000
50%             43.230000          43.070000       0.000000
75%            107.200000         106.990000      11.780000
max         280102.230000       44433.550000    4932.760000

View data types of the dataset

In [24]:
data.dtypes

TRANS DATE            datetime64[ns]
TRANS VAT DESC                object
ORIGINAL GROSS AMT           float64
ORIGINAL CUR                  object
BILLING GROSS AMT            float64
BILLING CUR CODE              object
MERCHANT NAME                 object
CARD NUMBER                   object
TRANS CAC CODE 1              object
TRANS CAC DESC 1              object
TRANS CAC CODE 2              object
TRANS CAC DESC 2              object
TRANS CAC CODE 3              object
Directorate                   object
TRANS TAX AMT                float64
BILLING CUR CODE.1            object
Directorates                  object
dtype: object

In [29]:
data.isnull().sum() #dataframe that contains null values
#the dataset indicates that in there are null values ​​in some columns

TRANS DATE                0
TRANS VAT DESC         4258
ORIGINAL GROSS AMT        0
ORIGINAL CUR              0
BILLING GROSS AMT         0
BILLING CUR CODE          0
MERCHANT NAME             0
CARD NUMBER               0
TRANS CAC CODE 1         46
TRANS CAC DESC 1         46
TRANS CAC CODE 2         89
TRANS CAC DESC 2         89
TRANS CAC CODE 3        127
Directorate            5393
TRANS TAX AMT         43991
BILLING CUR CODE.1    43991
Directorates          43934
dtype: int64

In [47]:
data['Directorate'].value_counts()

CYP&F SCHOOLS                   20516
Local Services                  12081
CYP&F                            4581
Corporate Resources              3416
Adult & Communities              1757
DEVELOPMENT                       533
Adult Social Care and Health      169
CYP&F                              67
CYP&F                              31
Development                        29
Local services                     10
Adults                              5
CYP&F Schools                       1
Name: Directorate, dtype: int64

In [48]:
data['ORIGINAL GROSS AMT'].value_counts()

15.00     337
10.00     302
20.00     246
3.00      222
50.00     216
         ... 
63.54       1
358.56      1
372.88      1
700.01      1
88.67       1
Name: ORIGINAL GROSS AMT, Length: 17493, dtype: int64

In [49]:
data['ORIGINAL CUR'].value_counts()

GBP    48294
EUR      123
USD      115
SEK        9
HUF        7
HKD        6
SGD        6
DKK        5
MYR        5
CNY        3
CZK        3
HRK        3
PLN        3
RON        2
NOK        2
QAR        1
CAD        1
PKR        1
Name: ORIGINAL CUR, dtype: int64

Analyzing the null data in the data set, it is decided to eliminate some of the variables by means of the drop function, that's because there are some columns that represent a very high percentage of nulls

In [50]:

data = data.drop(['TRANS TAX AMT','BILLING CUR CODE.1','Directorates'], axis = 1) #drop columns with large number of missing values


In [51]:
data

TRANS DATE TRANS VAT DESC  ...  TRANS CAC CODE 3          Directorate
0     2017-01-03             VR  ...               A00  Adult & Communities
1     2017-01-24             VR  ...               A00  Adult & Communities
2     2017-02-01             VR  ...               A00  Adult & Communities
3     2017-01-04             VR  ...               A00  Adult & Communities
4     2017-01-05             VR  ...               A00  Adult & Communities
...          ...            ...  ...               ...                  ...
48584 2017-11-23             VR  ...               A00       Local Services
48585 2017-11-17             VR  ...               A00       Local Services
48586 2017-11-17             VR  ...               A00       Local Services
48587 2017-11-24             VR  ...               A00       Local Services
48588 2017-11-24             VZ  ...               A00       Local Services

[48589 rows x 14 columns]

The other empty values ​​it is decided to impute or fill values ​​with the average value of the data

In [57]:
#Imput null data for '0' in the categorical variables because after numerical codes are added to the cagegoric variables
NAN_SUBSTITUTION_VALUE = "0"
data = data.astype('string')
data = data.fillna(NAN_SUBSTITUTION_VALUE)
data = data.astype('category')

In [58]:
data

TRANS DATE TRANS VAT DESC  ... TRANS CAC CODE 3          Directorate
0      2017-01-03             VR  ...              A00  Adult & Communities
1      2017-01-24             VR  ...              A00  Adult & Communities
2      2017-02-01             VR  ...              A00  Adult & Communities
3      2017-01-04             VR  ...              A00  Adult & Communities
4      2017-01-05             VR  ...              A00  Adult & Communities
...           ...            ...  ...              ...                  ...
48584  2017-11-23             VR  ...              A00       Local Services
48585  2017-11-17             VR  ...              A00       Local Services
48586  2017-11-17             VR  ...              A00       Local Services
48587  2017-11-24             VR  ...              A00       Local Services
48588  2017-11-24             VZ  ...              A00       Local Services

[48589 rows x 14 columns]

In [59]:
data.isnull().sum() #view null data 

TRANS DATE            0
TRANS VAT DESC        0
ORIGINAL GROSS AMT    0
ORIGINAL CUR          0
BILLING GROSS AMT     0
BILLING CUR CODE      0
MERCHANT NAME         0
CARD NUMBER           0
TRANS CAC CODE 1      0
TRANS CAC DESC 1      0
TRANS CAC CODE 2      0
TRANS CAC DESC 2      0
TRANS CAC CODE 3      0
Directorate           0
dtype: int64

The dataset indicates that in there are **null values ​​in some columns,** for example in TRANS VAT DESC, TRANS CAC CODE 1,TRANS CAC DESC 1,TRANS CAC CODE 2,TRANS CAC DESC 2,TRANS CAC CODE 3,Directorate,TRANS TAX AMT,BILLING CUR CODE.1,
Directorates.